pkg/providers/vsphere/reconciler/reconciler.go

```diff
@@ -93,201 +71,74 @@ func SetupEnvVars(ctx context.Context, vsphereDatacenter *anywherev1.VSphereData
-		if !conditions.Has(capiCluster, managedEtcdReadyCondition) || conditions.IsFalse(capiCluster, managedEtcdReadyCondition) {
-			log.Info("Waiting for etcd to be ready", "cluster", cluster.Name)
-			return controller.Result{Result: &ctrl.Result{
-				RequeueAfter: defaultRequeueTime,
-			}}, nil
+	if !dataCenterConfig.Status.SpecValid {
+		failureMessage := fmt.Sprintf("VSphereDatacenterConfig %s is invalid", dataCenterConfig.Name)
+		if dataCenterConfig.Status.FailureMessage != nil {
+			failureMessage += ": " + *dataCenterConfig.Status.FailureMessage
 		}
-	}
-
-	if !conditions.IsTrue(capiCluster, controlPlaneReadyCondition) {
-		log.Info("waiting for control plane to be ready", "cluster", capiCluster.Name, "kind", capiCluster.Kind)
-		return controller.Result{Result: &ctrl.Result{
-			RequeueAfter: defaultRequeueTime,
-		}}, err
-	}
 
-	if result, err := r.reconcileCNI(ctx, cluster, capiCluster, specWithBundles, log); err != nil {
-		return result, err
+		log.Error(nil, failureMessage)
+		clusterSpec.Cluster.Status.FailureMessage = &failureMessage
+		return controller.Result{}, nil
 	}
-
 	return controller.Result{}, nil
 }
 
-func (r *Reconciler) reconcileCNI(ctx context.Context, cluster *anywherev1.Cluster, capiCluster *clusterv1.Cluster, specWithBundles *c.Spec, log logr.Logger) (controller.Result, error) {
-	if !conditions.Has(cluster, cniSpecAppliedCondition) || conditions.IsFalse(capiCluster, cniSpecAppliedCondition) {
-		log.Info("Getting remote client", "client for cluster", capiCluster.Name)
-		key := client.ObjectKey{
-			Namespace: capiCluster.Namespace,
-			Name:      capiCluster.Name,
-		}
-		remoteClient, err := r.tracker.GetClient(ctx, key)
-		if err != nil {
-			return controller.Result{}, err
-		}
-
-		log.Info("About to apply CNI")
+func (r *Reconciler) ValidateMachineConfigs(ctx context.Context, log logr.Logger, clusterSpec *c.Spec) (controller.Result, error) {
+	log = log.WithValues("phase", "validateMachineConfigs")
+	machineConfigs := clusterSpec.VSphereMachineConfigs
+	datacenterConfig := clusterSpec.VSphereDatacenter
 
-		helm := executables.NewHelm(executables.NewExecutable("helm"), executables.WithInsecure())
-		cilium := cilium.NewCilium(nil, helm)
+	vsphereClusterSpec := vsphere.NewSpec(clusterSpec, machineConfigs, datacenterConfig)
 
-		if err != nil {
-			return controller.Result{}, err
-		}
-		ciliumSpec, err := cilium.GenerateManifest(ctx, specWithBundles, []string{constants.CapvSystemNamespace})
-		if err != nil {
-			return controller.Result{}, err
-		}
-		if err := serverside.ReconcileYaml(ctx, remoteClient, ciliumSpec); err != nil {
-			return controller.Result{}, err
-		}
-		conditions.MarkTrue(cluster, cniSpecAppliedCondition)
+	if err := r.validator.ValidateClusterMachineConfigs(ctx, vsphereClusterSpec); err != nil {
+		failureMessage := fmt.Sprintf("VSphereMachineConfig is invalid: %v", err)
+		log.Error(nil, failureMessage)
+		clusterSpec.Cluster.Status.FailureMessage = &failureMessage
```

Comment:
are we testing all these stuff?

Reply:
yes, it's tested in [this file](https://github.com/aws/eks-anywhere/blob/main/controllers/cluster_controller_test.go) & I'm going to add more tests in my next PR that will include a `reconciler_test.go` file for vsphere reconciler

Reply:
will be sure to have tests with assertions to cover these cases!